python/rmm/rmm/_lib/device_buffer.pyx

```diff
@@ -314,7 +314,7 @@ cdef class DeviceBuffer:
         cdef size_t s = dbp.size()
 
         cdef bytes b = PyBytes_FromStringAndSize(NULL, s)
-        cdef unsigned char* p = <unsigned char*>PyBytes_AS_STRING(b)
+        cdef unsigned char* p = <unsigned char*>PyBytes_AsString(b)
```

Comment:
```suggestion
        cdef unsigned char* p = b
```

Reply:
Do you know what this code actually does? Without having Cython generate the code I wouldn't know offhand if this translates into the right operation in Cython.

Reply:
Yes checked locally. It does exactly the same thing

Here is the logic in Cython. If the limited API is used, it [sets a bunch of `#define` constants]( https://github.com/cython/cython/blob/884ce74ae2229ee4191e96751635243f333fa143/Cython/Utility/ModuleSetupCode.c#L190 ). One of which is to disable [all CPython macros]( https://github.com/cython/cython/blob/884ce74ae2229ee4191e96751635243f333fa143/Cython/Utility/ModuleSetupCode.c#L227-L228 ), which means it uses [`PyBytes_AsString` with appropriate type casting]( https://github.com/cython/cython/blob/884ce74ae2229ee4191e96751635243f333fa143/Cython/Utility/TypeConversion.c#L70-L76 ). This particular case [uses `__Pyx_PyBytes_AsWritableUString`]( https://github.com/cython/cython/blob/884ce74ae2229ee4191e96751635243f333fa143/Cython/Utility/TypeConversion.c#L73 )

We can demonstrate this with code like this

<details>

```cython
from cpython.bytes cimport PyBytes_FromStringAndSize

cdef func():
    cdef ssize_t s = 10
    cdef bytes b = PyBytes_FromStringAndSize(NULL, s)
    cdef unsigned char* p = <unsigned char*>b
```

```c
  /* "_cython_magic_ea40401c76968889e5d6a433b8b09d4c162e5df1.pyx":5
 * 
 * cdef func():
 *     cdef ssize_t s = 10             # <<<<<<<<<<<<<<
 *     cdef bytes b = PyBytes_FromStringAndSize(NULL, s)
 *     cdef unsigned char* p = b
 */
  __pyx_v_s = 10;

  /* "_cython_magic_ea40401c76968889e5d6a433b8b09d4c162e5df1.pyx":6
 * cdef func():
 *     cdef ssize_t s = 10
 *     cdef bytes b = PyBytes_FromStringAndSize(NULL, s)             # <<<<<<<<<<<<<<
 *     cdef unsigned char* p = b
 * 
 */
  __pyx_t_1 = PyBytes_FromStringAndSize(NULL, __pyx_v_s); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 6, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_b = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "_cython_magic_ea40401c76968889e5d6a433b8b09d4c162e5df1.pyx":7
 *     cdef ssize_t s = 10
 *     cdef bytes b = PyBytes_FromStringAndSize(NULL, s)
 *     cdef unsigned char* p = b             # <<<<<<<<<<<<<<
 * 
 * 
 */
  if (unlikely(__pyx_v_b == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "expected bytes, NoneType found");
    __PYX_ERR(0, 7, __pyx_L1_error)
  }
  __pyx_t_2 = __Pyx_PyBytes_AsWritableUString(__pyx_v_b); if (unlikely((!__pyx_t_2) && PyErr_Occurred())) __PYX_ERR(0, 7, __pyx_L1_error)
  __pyx_v_p = __pyx_t_2;
```

</details>

Cython has an added `None` check, but this is a pointer comparison in C. One could probably disable this with the right Cython directive

Reply:
Though TBH if we feel the need to do this much deep-diving on relatively simple Cython codes to guarantee limited/stable API is used, it just may not be ready to rely on

Reply:
Sorry, I should have been clearer. I trust that if this compiled, it wouldn't be using anything that's not in the limited API. I'm familiar with the macros etc that you're referring to. What I didn't know without testing was whether Cython would transparently translate a `bytes` object into an `unsigned char *`. IOW my question wasn't related to whether Cython would do the right thing, it was regarding whether what you wrote is valid Cython code. 

As far as trusting cython's limited API support, we definitely aren't there yet. I would suggest reading https://github.com/rapidsai/build-planning/issues/42 a bit more closely. I included some notes to the effect that Cython doesn't really have official support for this yet and that I'm using development builds. My goal at the moment is to see what it would take to make RAPIDS code compatible with the limited API, in a world where Cython support is available and reliable. Outside of this change, I think all of the other changes in RAPIDS (there aren't many) don't have any adverse performance effects whatsoever. This rmm change is the only case where we might lose some perf, although I am still somewhat skeptical that this is actually in the critical code path for any performance-critical segments so I think it's worth the tradeoff to make our code limited API-compatible and easy to test with newer versions of Cython that will eventually support the limited API.

Reply:
Ok yes it is valid Cython 🙂

There are a full variety of conversions like this that Cython can do. Here are [a few relevant ones like this]( https://github.com/cython/cython/blob/8d5eb48473db9fda485c215683a6f22d1bb91c08/tests/errors/string_assignments.pyx#L12-L16 )

Reply:
Cool yeah I wasn't sure if this was one that Cython would handle automatically. Good to know.